### DISCUSSION TITLE
ESM imports in the client

### DISCUSSION THREAD

**Discussion by vicary** [vercel/next.js#45180] [Ideas]
### Describe the feature you'd like to request

We would like to port our existing plugin system, which dynamically imports browser compatible ESM modules via `await import(moduleName);`, into our Next.js app.

Is it possible for the Next.js Compiler to skip some of the dynamic import statements, preserving it as-is so it runs at client side?

### Describe the solution you'd like

There are a few ways to approach this,

1. Expose `ignoreDynamics` from SWC.
    I know this may face whatever reason `next/dynamic` exists in the first place, listing it out just in case it is not.
2. Create a new package, e.g. `next/import`, which transforms into the original `import()` statement in the resulting bundle.

#### External Web Workers

The only known option to us is to setup a separate compiler and build our plugins into Web Worker scripts, which is then served in the `/public` directory for our current Next.js app to use.

There are a few downsides using this alternative,
1. Worker IPC introduces a new class of error handling and timeout management.
2. The maintenance of a separated set of compiler pipeline partially defeats the purpose of using Next.js in the first place.
3. Next.js and React packages usually doesn't work in Web Worker environment, thus any JSX related logics must be further decoupled back into the "main thread" for rendering. This introduces a new learning curve for our team.

**Comment by GabenGar** [comment]
Wouldn't dynamic keys for dynamic imports bail the bundler out from resolving them at build time?

**Comment by vicary** [comment]
@GabenGar Unfortunately the `import(moduleName)` keyword will become something like `__webpackRequire(moduleName)`.

**Comment by GabenGar** [comment]
By `moduleName` do you mean a string literal or a non-const string variable?

**Comment by vicary** [comment]
It's a variable.
